fuel/datasets/base.py

```diff
@@ -177,88 +190,142 @@ def filter_sources(self, data):
-        This allows datasets to be pickled and unpickled, even in
-        environments where the original data is unavailable or has changed
-        position.
+    @property
+    def num_examples(self):
+        try:
+            num_examples, = set(len(iterable) for iterable in self.iterables)
+            return num_examples
+        except TypeError:
+            return float('nan')
 
-        """
-        pass
+    def open(self):
+        iterators = [_iter(channel) for channel in self.iterables]
+        return izip(*iterators)
 
+    def get_data(self, state=None, request=None):
+        if state is None or request is not None:
+            raise ValueError
+        return next(state)
 
 
+class IndexableDataset(Dataset):
+    """Creates a dataset from a set of indexable containers.
+
+    Parameters
+    ----------
+    indexables : :class:`~collections.OrderedDict` or indexable
+        The indexable(s) to provide interface to. This means it must
+        support the syntax ```indexable[0]``. If an
+        :class:`~collections.OrderedDict` is given, its values should be
+        indexables providing data, and its keys strings that are used as
+        source names. If a single indexable is given, it will be given the
+        source ``data``.
+
+    Attributes
+    ----------
+    indexables : list
+        A list of indexable objects.
+
+    Notes
+    -----
+    If the indexable data is very large, you might want to consider using
+    the :func:`.do_not_pickle_attributes` decorator to make sure the data
+    doesn't get pickled with the dataset, but gets reloaded/recreated
+    instead.
+
+    This dataset also uses the source names to create properties that
+    provide easy access to the data.
+
+    """
+    def __init__(self, indexables, start=None, stop=None, **kwargs):
+        if isinstance(indexables, dict):
+            self.provides_sources = tuple(indexables.keys())
+        else:
+            self.provides_sources = ('data',)
+        super(IndexableDataset, self).__init__(**kwargs)
+        if isinstance(indexables, dict):
+            self.indexables = [indexables[source][start:stop]
+                               for source in self.sources]
+            if not all(len(indexable) == len(self.indexables[0])
+                       for indexable in self.indexables):
+                raise ValueError("sources have different lengths")
+        else:
+            self.indexables = [indexables]
+
+        self.example_iteration_scheme = SequentialExampleScheme(
+            self.num_examples)
+
+        def property_factory(source):
+            def property_(self):
+                if source not in self.sources:
+                    raise AttributeError
+                return self.indexables[self.sources.index(source)]
+            return property_
+        for source in self.sources:
+            setattr(self.__class__, source, property(property_factory(source)))
```

Comment:
The `provides_sources` isn't defined until after the call to `__init__` (see lines 296, 298) so a metaclass would have to attach the properties to the class at exactly the same point as I do now (i.e. after calling `__init__` from the metaclass's `__call__` method). So in effect you would do *exactly* the same thing, the only difference being that the code is hidden in a metaclass instead of being accessible to users in the code.